crates/jstz_core/src/kv/transaction.rs

```diff
@@ -471,116 +560,178 @@ impl<'a, V> Entry<'a, V> {
+        let mut inner = self.guard.borrow_mut();
+        GuardedMut::new(
+            self.guard.clone(),
+            inner.stack[self.snapshot_idx]
+                .lookup_mut(&self.key)
+                .unwrap() //TODO: Add errors
+                .as_mut::<V>()
+                .unwrap(),
+        )
     }
 
     /// Convert the entry into a mutable reference to its value.
-    pub fn into_mut(self) -> &'a mut V
+    pub fn into_mut(self) -> GuardedMut<'a, V>
     where
         V: Value,
     {
-        self.inner
-            .into_mut()
-            .as_mut()
-            .expect("Invalid type id invariant")
+        let mut inner = self.guard.borrow_mut();
+        GuardedMut::new(
+            self.guard.clone(),
+            inner.stack[self.snapshot_idx]
+                .lookup_mut(&self.key)
+                .unwrap()
+                .as_mut::<V>()
+                .unwrap(),
+        )
     }
 
     /// Sets the value of the entry and returns the entry's old value.
     pub fn insert(&mut self, value: V) -> V
     where
         V: Value,
     {
-        std::mem::replace(self.get_mut(), value)
+        std::mem::replace(&mut self.get_mut(), value)
     }
+}
 
-    /// Take the value of the entry out of the snapshot, and return it.
-    pub fn remove(self) -> V
-    where
-        V: Value,
-    {
-        self.remove_edits.insert(self.key().clone());
-        self.inner
-            .remove()
-            .into_value()
-            .expect("Invalid type id invariant")
+// Guarded<T>/GuardedMut<T> – keeps the mutex guard alive while reference is alive
+
+pub struct Guarded<'a, T: ?Sized + 'a> {
+    value: *const T,
+    pub _guard: RcGuardInner,
+    _marker: PhantomData<&'a ()>,
+}
+
+impl<'a, T: ?Sized + 'a> Guarded<'a, T> {
+    pub fn new(guard: RcGuardInner, value: &T) -> Guarded<'a, T> {
+        Guarded {
+            value,
+            _guard: guard,
+            _marker: PhantomData,
+        }
+    }
+    pub fn clone_guard(&self) -> RcGuardInner {
+        self._guard.clone()
+    }
+}
+
+impl<'a, T: ?Sized + 'a> std::ops::Deref for Guarded<'a, T> {
+    type Target = T;
+    fn deref(&self) -> &'a Self::Target {
+        unsafe { &*self.value }
+    }
+}
+
+pub struct GuardedMut<'a, T: ?Sized + 'a> {
+    value: *mut T,
+    pub _guard: RcGuardInner,
+    _marker: PhantomData<&'a ()>,
+}
+
+impl<'a, T: ?Sized + 'a> GuardedMut<'a, T> {
+    pub fn new<'b>(guard: RcGuardInner, value: &mut T) -> GuardedMut<'b, T> {
+        GuardedMut {
+            value,
+            _guard: guard,
+            _marker: PhantomData,
+        }
+    }
+    pub fn clone_guard(&self) -> RcGuardInner {
+        self._guard.clone()
+    }
+}
+
+impl<'a, T: ?Sized + 'a> std::ops::Deref for GuardedMut<'a, T> {
+    type Target = T;
+    fn deref(&self) -> &'a Self::Target {
+        unsafe { &*self.value }
+    }
+}
+
+impl<'a, T: ?Sized + 'a> std::ops::DerefMut for GuardedMut<'a, T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
```

Comment:
Although the `'a` lifetime should transitively applied to `&mut Self::Target` since `'a` lives at least as long as `GuardedMut`, I think its better to make the relationship explicit.

```suggestion
    fn deref_mut(&mut self) -> &'a mut Self::Target {
```